From: Jiri Kosina <jkosina@suse.cz>
Subject: PIE executable randomization
References: FATE120276

Use mmap()'s randomization functionality in such a way that it maps the
main executable of (specially compiled/linked -pie/-fpie) ET_DYN binaries
onto a random address (in cases in which mmap() is allowed to perform a
randomization).

[akpm@linux-foundation.org: fix used-uninitialsied warning]
[kamezawa.hiroyu@jp.fujitsu.com: fixed ia32 ELF on x86_64 handling]
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
Cc: KAMEZAWA Hiroyuki <kamezawa.hiroyu@jp.fujitsu.com>
Cc: Arjan van de Ven <arjan@infradead.org>
Cc: Ingo Molnar <mingo@elte.hu>
Cc: Roland McGrath <roland@redhat.com>
Cc: Jakub Jelinek <jakub@redhat.com>
Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: "Luck, Tony" <tony.luck@intel.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Automatically created from "patches.suse/aslr-pie-executable-randomization.patch" by xen-port-patches.py

Index: head-2007-11-30/include/asm-x86/mach-xen/asm/pgtable_64.h
===================================================================
--- head-2007-11-30.orig/include/asm-x86/mach-xen/asm/pgtable_64.h	2007-12-04 08:34:05.000000000 +0100
+++ head-2007-11-30/include/asm-x86/mach-xen/asm/pgtable_64.h	2007-12-04 13:59:24.000000000 +0100
@@ -539,6 +539,7 @@ pte_t *lookup_address(unsigned long addr
 		direct_remap_pfn_range(vma,vaddr,pfn,size,prot,DOMID_IO)
 
 #define HAVE_ARCH_UNMAPPED_AREA
+#define HAVE_ARCH_UNMAPPED_AREA_TOPDOWN
 
 #define pgtable_cache_init()   do { } while (0)
 #define check_pgt_cache()      do { } while (0)
